Use attributes for `dangling_pointers_from_temporaries` lint

gavincrawford opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

https://github.com/rust-lang/rust/issues/132281

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

r? @Urgau
-->
<!-- homu-ignore:end -->
Checking for dangling pointers by function name isn't ideal, and leaves out certain pointer-returning methods that don't follow the `as_ptr` naming convention. Using an attribute for this lint cleans things up and allows more thorough coverage of other methods, such as `UnsafeCell::get()`.

gavincrawford committed:
Add #[rustc_as_ptr] attribute

gavincrawford committed:
Tag relevant functions with #[rustc_as_ptr] attribute

gavincrawford committed:
Check for #[rustc_as_ptr] attribute instead of searching for function names

Co-authored-by: Urgau <3616612+Urgau@users.noreply.github.com>

GrigorenkoPV commented on compiler/rustc_feature/src/builtin_attrs.rs:
```diff
@@ -875,6 +875,11 @@ pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
+    rustc_attr!(
+        rustc_as_ptr, Normal, template!(Word), ErrorFollowing,
+        EncodeCrossCrate::Yes,
+        "#[rustc_as_ptr] is used to mark functions that implement `as_ptr` or `as_ptr_mut`."
```
Well, "implement" is usually refers to traits in Rust. I guess maybe it should say something about providing a pointer to the inner allocation instead.

gavincrawford replied:
Good catch, that could end up being confusing. Just fixed

GrigorenkoPV commented on library/core/src/cell.rs:
```diff
@@ -2159,6 +2159,7 @@ impl<T: ?Sized> UnsafeCell<T> {
     #[inline(always)]
     #[stable(feature = "rust1", since = "1.0.0")]
     #[rustc_const_stable(feature = "const_unsafecell_get", since = "1.32.0")]
+    #[cfg_attr(not(bootstrap), rustc_as_ptr)]
```
There's also `SyncUnsafeCell::get` below, which should also be marked I think.

gavincrawford replied:
Thanks. Went ahead and tagged that one as well as its raw pointer counterpart

GrigorenkoPV reviewed (changes requested):
Could you also please add test cases for the newly marked methods to `tests/ui/lint/dangling-pointers-from-temporaries/types.rs`?

GrigorenkoPV commented:
Other than the 3 points above, LGTM.

[Attribute validation](https://github.com/rust-lang/rust/pull/132732/files#diff-77c442894af49f492927d99f7e60f4365a4829100c0c52f565f42dc184712c76R165-R167) turned out to be way simpler than I expected, which is a pleasant surprise.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11727291994/job/32668028365) [(plain)](https://github.com/rust-lang/rust/commit/85c93188a263ce397e5ec626afcd7d62eda02ef3/checks/32668028365/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/allow.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/allow.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/allow" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/cstring-as-ptr.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/cstring-as-ptr.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/cstring-as-ptr.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/cstring-as-ptr" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- stderr -------------------------------
warning: lint `temporary_cstring_as_ptr` has been renamed to `dangling_pointers_from_temporaries`
##[warning]  --> /checkout/tests/ui/lint/dangling-pointers-from-temporaries/cstring-as-ptr.rs:2:9
   |
---
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/example-from-issue123613.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/example-from-issue123613.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/example-from-issue123613" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/calls.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/calls.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/calls.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/calls" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/ext.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/ext.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/ext.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/ext" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/methods.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/methods.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/methods.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/methods" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/types.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/types.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/types.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/types" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/temporaries.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/temporaries.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/temporaries.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/temporaries" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none



```

</details>

Urgau commented on compiler/rustc_lint/src/dangling.rs:
```diff
@@ -133,10 +133,11 @@ impl DanglingPointerSearcher<'_, '_> {
 fn lint_expr(cx: &LateContext<'_>, expr: &Expr<'_>) {
     if let ExprKind::MethodCall(method, receiver, _args, _span) = expr.kind
-        && matches!(method.ident.name, sym::as_ptr | sym::as_mut_ptr)
+        && let Some(fn_id) = cx.typeck_results().type_dependent_def_id(method.hir_id)
```
The hir id of `method` won't resolve to the function, for that you should use the one from the method call it-self, in this case from `expr`.
```suggestion
        && let Some(fn_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id)
```

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11729660528/job/32675830243) [(plain)](https://github.com/rust-lang/rust/commit/77fd84ab62957cd96688b37b6946b0b4bcd10dbd/checks/32675830243/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/allow.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/allow.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/allow" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/cstring-as-ptr.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/cstring-as-ptr.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/cstring-as-ptr.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/cstring-as-ptr" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- stderr -------------------------------
warning: lint `temporary_cstring_as_ptr` has been renamed to `dangling_pointers_from_temporaries`
##[warning]  --> /checkout/tests/ui/lint/dangling-pointers-from-temporaries/cstring-as-ptr.rs:2:9
   |
---
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/example-from-issue123613.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/example-from-issue123613.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/example-from-issue123613" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/calls.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/calls.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/calls.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/calls" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/ext.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/ext.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/ext.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/ext" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/methods.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/methods.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/methods.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/methods" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/temporaries.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/temporaries.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/temporaries.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/temporaries" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none


---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/types.rs stdout ----
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/types.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/types.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/types" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stderr: none



```

</details>

GrigorenkoPV commented on library/core/src/cell.rs:
```diff
@@ -2323,6 +2324,7 @@ impl<T: ?Sized> SyncUnsafeCell<T> {
     /// Gets a mutable pointer to the wrapped value.
     ///
     /// See [`UnsafeCell::get`] for details.
+    #[cfg_attr(not(bootstrap), rustc_as_ptr)]
```
We don't actually need this one, as it
1. Just like with `MaybeUninit::slice_as_ptr`, it is not callable in `temporary.method()` form.
2. It already takes a pointer as an argument, while `dangling_pointers_from_temporaries` is all about converting from a temporary to a pointer, not pointer-to-pointer casts.

GrigorenkoPV replied:
To clarify, the lint is about crossing the boundary between borrow-checked lifetime-having Rust types and the unmanaged unchecked untracked land of raw pointers. And in particular, the specific case where "the temporary will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned".

This function does not cross the said boundary.

gavincrawford replied:
Sorry, I see the issue. Just pushed a fix.

Just to confirm my understanding, the reason this does not apply is because the `raw_get` function simply casts a constant pointer to a mutable one, and thus does not concern the crossing of the managed-unmanaged boundary, as opposed to the `get` function that returns a raw pointer to the inner value?

GrigorenkoPV replied:
Yes

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11735764009/job/32693936413) [(plain)](https://github.com/rust-lang/rust/commit/05fb23c5dbe4affc47ef7b020c81bd2d11071895/checks/32693936413/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/lint/dangling-pointers-from-temporaries/types.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/types/types.stderr"
diff of stderr:

47    = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<CString>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned
49 
- error: a dangling pointer will be produced because the temporary `Box<[u8]>` will be dropped
-   --> $DIR/types.rs:28:28
-    |
-    |
- LL |     declval::<Box<[u8]>>().as_ptr();
-    |     ---------------------- ^^^^^^ this pointer will immediately be invalid
-    |     this `Box<[u8]>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
-    |
-    |
-    = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<[u8]>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned
- 
61 error: a dangling pointer will be produced because the temporary `Box<str>` will be dropped
62   --> $DIR/types.rs:30:27
63    |
63    |

80    = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<CStr>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned
82 
82 
- error: a dangling pointer will be produced because the temporary `[u8; 10]` will be dropped
-    |
- LL |     declval::<[u8; 10]>().as_ptr();
-    |     --------------------- ^^^^^^ this pointer will immediately be invalid
-    |     |
-    |     |
-    |     this `[u8; 10]` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
-    |
-    = note: pointers do not have a lifetime; when calling `as_ptr` the `[u8; 10]` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned
- 
- error: a dangling pointer will be produced because the temporary `Box<[u8; 10]>` will be dropped
-   --> $DIR/types.rs:36:32
-    |
-    |
- LL |     declval::<Box<[u8; 10]>>().as_ptr();
-    |     -------------------------- ^^^^^^ this pointer will immediately be invalid
-    |     this `Box<[u8; 10]>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
-    |
-    |
-    = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<[u8; 10]>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned
- 
- 
105 error: a dangling pointer will be produced because the temporary `Box<Vec<u8>>` will be dropped
107    |


124    = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<String>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned
126 
126 
- error: a dangling pointer will be produced because the temporary `Box<Box<Box<Box<[u8]>>>>` will be dropped
-    |
-    |
- LL |     declval::<Box<Box<Box<Box<[u8]>>>>>().as_ptr();
-    |     ------------------------------------- ^^^^^^ this pointer will immediately be invalid
-    |     |
-    |     this `Box<Box<Box<Box<[u8]>>>>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
-    |
-    = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<Box<Box<Box<[u8]>>>>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned
- 
- 
138 error: a dangling pointer will be produced because the temporary `Cell<u8>` will be dropped
140    |


168    = note: pointers do not have a lifetime; when calling `as_ptr` the `Vec<AsPtrFake>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned
170 
- error: aborting due to 15 previous errors
+ error: aborting due to 11 previous errors
172 
---
To only update this specific test, also pass `--test-args lint/dangling-pointers-from-temporaries/types.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/dangling-pointers-from-temporaries/types.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/dangling-pointers-from-temporaries/types" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- stderr -------------------------------
error: a dangling pointer will be produced because the temporary `CString` will be dropped
##[error]  --> /checkout/tests/ui/lint/dangling-pointers-from-temporaries/types.rs:20:26
   |
   |
LL |     declval::<CString>().as_ptr();
   |     -------------------- ^^^^^^ this pointer will immediately be invalid
   |     this `CString` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `CString` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned
note: the lint level is defined here
  --> /checkout/tests/ui/lint/dangling-pointers-from-temporaries/types.rs:1:9
   |
LL | #![deny(dangling_pointers_from_temporaries)]
LL | #![deny(dangling_pointers_from_temporaries)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: a dangling pointer will be produced because the temporary `String` will be dropped
##[error]  --> /checkout/tests/ui/lint/dangling-pointers-from-temporaries/types.rs:22:25
   |
LL |     declval::<String>().as_ptr();
   |     ------------------- ^^^^^^ this pointer will immediately be invalid
   |     this `String` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `String` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned


error: a dangling pointer will be produced because the temporary `Vec<u8>` will be dropped
   |
   |
LL |     declval::<Vec<u8>>().as_ptr();
   |     -------------------- ^^^^^^ this pointer will immediately be invalid
   |     this `Vec<u8>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `Vec<u8>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned


error: a dangling pointer will be produced because the temporary `Box<CString>` will be dropped
   |
   |
LL |     declval::<Box<CString>>().as_ptr();
   |     ------------------------- ^^^^^^ this pointer will immediately be invalid
   |     |
   |     this `Box<CString>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<CString>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned

error: a dangling pointer will be produced because the temporary `Box<str>` will be dropped
##[error]  --> /checkout/tests/ui/lint/dangling-pointers-from-temporaries/types.rs:30:27
   |
   |
LL |     declval::<Box<str>>().as_ptr();
   |     --------------------- ^^^^^^ this pointer will immediately be invalid
   |     |
   |     this `Box<str>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<str>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned


error: a dangling pointer will be produced because the temporary `Box<CStr>` will be dropped
   |
   |
LL |     declval::<Box<CStr>>().as_ptr();
   |     ---------------------- ^^^^^^ this pointer will immediately be invalid
   |     this `Box<CStr>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<CStr>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned


error: a dangling pointer will be produced because the temporary `Box<Vec<u8>>` will be dropped
   |
   |
LL |     declval::<Box<Vec<u8>>>().as_ptr();
   |     ------------------------- ^^^^^^ this pointer will immediately be invalid
   |     this `Box<Vec<u8>>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<Vec<u8>>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned


error: a dangling pointer will be produced because the temporary `Box<String>` will be dropped
   |
LL |     declval::<Box<String>>().as_ptr();
   |     ------------------------ ^^^^^^ this pointer will immediately be invalid
   |     |
   |     |
   |     this `Box<String>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `Box<String>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned


error: a dangling pointer will be produced because the temporary `Cell<u8>` will be dropped
   |
   |
LL |     declval::<Cell<u8>>().as_ptr();
   |     --------------------- ^^^^^^ this pointer will immediately be invalid
   |     this `Cell<u8>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `Cell<u8>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned

error: a dangling pointer will be produced because the temporary `MaybeUninit<u8>` will be dropped
##[error]  --> /checkout/tests/ui/lint/dangling-pointers-from-temporaries/types.rs:46:34
   |
   |
LL |     declval::<MaybeUninit<u8>>().as_ptr();
   |     ---------------------------- ^^^^^^ this pointer will immediately be invalid
   |     this `MaybeUninit<u8>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `MaybeUninit<u8>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned


error: a dangling pointer will be produced because the temporary `Vec<AsPtrFake>` will be dropped
   |
   |
LL |     declval::<Vec<AsPtrFake>>().as_ptr();
   |     --------------------------- ^^^^^^ this pointer will immediately be invalid
   |     |
   |     this `Vec<AsPtrFake>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
   |
   = note: pointers do not have a lifetime; when calling `as_ptr` the `Vec<AsPtrFake>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned

error: aborting due to 11 previous errors
------------------------------------------

```

</details>

Urgau commented:
@gavincrawford Could you add the missing `#[rustc_as_ptr]` to the slice methods, so that the tests passes.

As well as doing @GrigorenkoPV [comment](https://github.com/rust-lang/rust/pull/132732#pullrequestreview-2421591171):

> Could you also please add test cases for the newly marked methods to `tests/ui/lint/dangling-pointers-from-temporaries/types.rs`?

To locally run the tests you can use `./x.py test --stage 1 tests/lint/dangling/`, don't hesitate to look at the [relevant chapter](https://rustc-dev-guide.rust-lang.org/tests/intro.html#compiletest) on the rustc-dev-guide for more guidance.

@rustbot author

gavincrawford commented:
Fixed the slice methods, working on the test cases, and had a quick question. In order for the lint to apply to `UnsafeCell`, we have to include it in this segment:
https://github.com/rust-lang/rust/blob/328b759142ddeae96da83176f103200009d3e3f1/compiler/rustc_lint/src/dangling.rs#L212-L218
Which works fine, and causes the lint to error out on `UnsafeCell::get()` as expected, but I'm struggling to find how to apply the same logic to `SyncUnsafeCell`. I understand that the synchronous version is just a wrapper around a normal `UnsafeCell`, and thus owns its inner allocation, but it does not have a definition that I can find in `LangItem`, nor a diagnostic name I can match against. Any ideas on how to handle this as of now?

GrigorenkoPV commented:
> but it does not have a definition that I can find in `LangItem`, nor a diagnostic name I can match against. Any ideas on how to handle this as of now?

I think you can make it into a new diagnostic item.

You will need to [mark the struct](https://github.com/rust-lang/rust/pull/128985/files#diff-75a679ddd1aae61c8b60e45cea1fb213a086caee3600993c68af9f7a09785e9eR307) & [add the new tag to a list of symbols](https://github.com/rust-lang/rust/pull/128985/files#diff-97beaf86f77cdca7e6b5583aa0764c8e0a493b957f6e5c1bb7ba40e15f2ed2d7R174).

Also you can probably remove `as_mut_ptr` & `as_ptr` from the list as well. The former is most likely used only in the lint that you have now made use attributes instead. IDK about the latter, though. It might be used somewhere else.

gavincrawford committed:
Update dangling pointer tests

gavincrawford committed:
Remove as_mut_ptr symbol

gavincrawford commented:
`as_ptr` had a couple other uses in other spots, so I left that one for now. All tests passed on my machine, so I guess let's just hope the CI agrees.

GrigorenkoPV commented on library/core/src/cell.rs:
```diff
@@ -2270,6 +2273,7 @@ impl<T: DispatchFromDyn<U>, U> DispatchFromDyn<UnsafeCell<U>> for UnsafeCell<T>
 /// See [`UnsafeCell`] for details.
 #[unstable(feature = "sync_unsafe_cell", issue = "95439")]
 #[repr(transparent)]
+#[rustc_diagnostic_item = "sync_unsafe_cell"]
```
The CI will most likely fail on this. For some dark magic reasons rustc wants this in the form of

```rust
#[cfg_attr(not(test), rustc_diagnostic_item = "Cell")]
```
[[src](https://github.com/rust-lang/rust/pull/132732/files#diff-75a679ddd1aae61c8b60e45cea1fb213a086caee3600993c68af9f7a09785e9eL307)]

Also I wonder if this should be in CamelCase instead or that doesn't matter too much.

gavincrawford replied:
The regular unsafe cell symbol is in there as `unsafe_cell`, so I was just following that as an example. I could change one, or both?

GrigorenkoPV commented on compiler/rustc_lint/src/dangling.rs:
```diff
@@ -201,22 +202,23 @@ fn is_temporary_rvalue(expr: &Expr<'_>) -> bool {
 
 // Array, Vec, String, CString, MaybeUninit, Cell, Box<[_]>, Box<str>, Box<CStr>,
```
Please, update this comment and, more importantly, the doc comment above that mentions that `[Sync]UnsafeCell::get()` methods are not accounted for.

GrigorenkoPV replied:
The doc comment on lines 46–48

gavincrawford commented:
> Okay, makes sense. I wonder where, probably will look at this later, as it sounds interesting.
It's used in a couple of Clippy lints, didn't look much more in to it then that.

> Once again, we'll hear from Urgau first (I do not have rights to approve PRs anyway), and they will probably ask you to squash the commits a bit, as that's [what's usually done here](https://github.com/rust-lang/rust/pull/128985#issuecomment-2417352647).
Yeah, I noticed the mess. Rebased and squashed the commits down.

As for the other stuff, `sync_unsafe_cell` will get CamelCase and I'll fix that comment as soon as I have some time to work on it.

gavincrawford committed:
Correct comments concerning updated dangling pointer lint

jieyouxu commented:
(Back ping to #132281)

Urgau commented:
@gavincrawford Is the PR ready for review? (I ask because it's still marked as waiting on author, do @`rustbot review` to change it)

gavincrawford commented:
Yes, sorry.

@rustbot review

Urgau commented on compiler/rustc_lint/src/dangling.rs:
```diff
@@ -199,24 +200,25 @@ fn is_temporary_rvalue(expr: &Expr<'_>) -> bool {
-fn is_interesting(tcx: TyCtxt<'_>, ty: Ty<'_>) -> bool {
+// Array, Vec, String, CString, MaybeUninit, Cell, Box<[_]>, Box<str>, Box<CStr>, UnsafeCell,
+// SyncUnsafeCell, or any of the above in arbitrary many nested Box'es.
+fn owns_allocation(tcx: TyCtxt<'_>, ty: Ty<'_>) -> bool {
```
With the addition of the attribute we should able to simplify this `owns_allocation` method to a simple check of the type and it's generic arguments.

Something like this would be a good starting point (*the check should probably be further refined*):
```rust
    match ty.kind() {
        ty::Array(_, _) => true,
        ty::Adt(_def, args) => args.iter().filter_map(|g| g.as_type()).all(|ty| !ty.is_any_ptr()),
        _ => false,
    }
```

GrigorenkoPV replied:
I think a `Box<Box<&[T]>>` would pass the check that you suggest, even though it should not

Urgau replied:
Yeah, we still need to take the derefs into account.

<details>

Probably something more like this:
```rust
    let adjs = cx.typeck_results().expr_adjustments(receiver);
    let origin_ty = cx.typeck_results().expr_ty(receiver);

    adjs.iter()
        .filter_map(|adj| match adj.kind {
            Adjust::Deref(_) => Some(adj.target),
            Adjust::NeverToAny
            | Adjust::Borrow(_)
            | Adjust::Pointer(_)
            | Adjust::ReborrowPin(_) => None,
        })
        .all(|ty| !ty.is_any_ptr())
        && match origin_ty.kind() {
            ty::Adt(_def, args) => {
                args.iter().filter_map(|g| g.as_type()).all(|ty| !ty.is_any_ptr())
            }
            _ => !origin_ty.is_any_ptr(),
        }
```

but, let's discussed that on the follow-up issue/pr.

</details>

Urgau commented:
LGTM. Let's fix the nit https://github.com/rust-lang/rust/pull/132732#discussion_r1845183216 and defer the simplification to a follow-up PR.

@rustbot author

gavincrawford commented:
Just fixed that comment. I squashed it with the other comment commit for clarity, as they're both just comment edits.

Urgau commented:
@bors r+

bors commented:
:pushpin: Commit 01fd384d5800deb1e96b0d8921e82317e9fcdb8b has been approved by `Urgau`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Urgau 01fd384d5800deb1e96b0d8921e82317e9fcdb8b -->
<!-- homu: {"type":"Approved","sha":"01fd384d5800deb1e96b0d8921e82317e9fcdb8b","approver":"Urgau","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Kobzol commented:
This PR seems to have caused a compile-time regression in the `nalgebra` crate (https://github.com/rust-lang/rust/pull/133234#issuecomment-2489036308). Looks like some queries are now called [more](https://perf.rust-lang.org/detailed-query.html?commit=0f0ea1170a9d1a2db1ae3684c1648810fe870814&benchmark=nalgebra-0.33.0-check&scenario=incr-patched%3A+println&base_commit=70e814bd9e532a302763f870c665c5af59c2b632&sort_idx=-11). If you have an idea on how to improve this, it would be nice to take a look. If not, I don't think that it's too terrible, and we can just eat the cost.
